utils/funtools/man/man1/funtable.1
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.32
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "funtable 1"
.TH funtable 1 "August 15, 2007" "version 1.4.0" "SAORD Documentation"
.SH "NAME"
\&\fBfuntable \- copy selected rows from a Funtools file to a FITS binary table\fR
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
funtable [\-a] [\-i|\-z] [\-m] [\-s cols] <iname> <oname> [columns]
.SH "OPTIONS"
.IX Header "OPTIONS"
.Vb 5
\&  -a    # append to existing output file as a table extension
\&  -i    # for image data, only generate X and Y columns
\&  -m    # for tables, write a separate file for each region
\&  -s "col1 ..." # columns on which to sort
\&  -z    # for image data, output zero-valued pixels
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBfuntable\fR selects rows from the specified
\&\s-1FITS\s0 Extension
(binary table only) of a \s-1FITS\s0 file, or from a non-FITS raw event
file, and writes those rows to a \s-1FITS\s0 binary table file. It also
will create a \s-1FITS\s0 binary table from an image or a raw array file.
.PP
The first argument to the program specifies the \s-1FITS\s0 file, raw event
file, or raw array file.  If \*(L"stdin\*(R" is specified, data are read from
the standard input. Use Funtools Bracket
Notation to specify \s-1FITS\s0 extensions, and filters.  The second
argument is the output \s-1FITS\s0 file.  If \*(L"stdout\*(R" is specified, the \s-1FITS\s0
binary table is written to the standard output.  By default, all
columns of the input file are copied to the output file.  Selected
columns can be output using an optional third argument in the form:
.PP
.Vb 1
\&  "column1 column1 ... columnN"
.Ve
.PP
The \fBfuntable\fR program generally is used to select rows from a
\&\s-1FITS\s0 binary table using
Table Filters
and/or
Spatial Region Filters.
For example, you can copy only selected rows (and output only selected
columns) by executing in a command such as:
.PP
.Vb 13
\&  [sh] funtable "test.ev[pha==1&&pi==10]" stdout "x y pi pha" | fundisp stdin
\&         X       Y     PHA        PI
\&   ------- ------- ------- ---------
\&         1      10       1        10
\&         1      10       1        10
\&         1      10       1        10
\&         1      10       1        10
\&         1      10       1        10
\&         1      10       1        10
\&         1      10       1        10
\&         1      10       1        10
\&         1      10       1        10
\&         1      10       1        10
.Ve
.PP
The special column \fB$REGION\fR can be specified to write the
region id of each row:
.PP
.Vb 12
\&  [sh $] funtable "test.ev[time-(int)time>=.99&&annulus(0 0 0 10 n=3)]" stdout 'x y time $REGION' | fundisp stdin
\&          X        Y                  TIME     REGION
\&   -------- -------- --------------------- ----------
\&          5       -6           40.99000000          3
\&          4       -5           59.99000000          2
\&         -1        0          154.99000000          1
\&         -2        1          168.99000000          1
\&         -3        2          183.99000000          2
\&         -4        3          199.99000000          2
\&         -5        4          216.99000000          2
\&         -6        5          234.99000000          3
\&         -7        6          253.99000000          3
.Ve
.PP
Here only rows with the proper fractional time and whose position also is
within one of the three annuli are written.
.PP
Columns can be excluded from display using a minus sign before the
column:
.PP
.Vb 12
\&  [sh $] funtable "test.ev[time-(int)time>=.99]" stdout "-time" | fundisp stdin
\&          X        Y      PHA         PI          DX          DY
\&   -------- -------- -------- ---------- ----------- -----------
\&          5       -6        5         -6        5.50       -6.50
\&          4       -5        4         -5        4.50       -5.50
\&         -1        0       -1          0       -1.50        0.50
\&         -2        1       -2          1       -2.50        1.50
\&         -3        2       -3          2       -3.50        2.50
\&         -4        3       -4          3       -4.50        3.50
\&         -5        4       -5          4       -5.50        4.50
\&         -6        5       -6          5       -6.50        5.50
\&         -7        6       -7          6       -7.50        6.50
.Ve
.PP
All columns except the time column are written.
.PP
In general, the rules for activating and de-activating columns are:
.IP "\(bu" 4
If only exclude columns are specified, then all columns but
the exclude columns will be activated.
.IP "\(bu" 4
If only include columns are specified, then only the specified columns
are activated.
.IP "\(bu" 4
If a mixture of include and exclude columns are specified, then
all but the exclude columns will be active; this last case
is ambiguous and the rule is arbitrary.
.PP
In addition to specifying columns names explicitly, the special
symbols \fI+\fR and \fI\-\fR can be used to activate and
de-activate \fIall\fR columns. This is useful if you want to
activate the \f(CW$REGION\fR column along with all other columns.  According
to the rules, the syntax \*(L"$REGION\*(R" only activates the region column
and de-activates the rest. Use \*(L"+ \f(CW$REGION\fR\*(R" to activate all
columns as well as the region column.
.PP
Ordinarily, only the selected table is copied to the output file.  In
a \s-1FITS\s0 binary table, it sometimes is desirable to copy all of the
other \s-1FITS\s0 extensions to the output file as well. This can be done by
appending a '+' sign to the name of the extension in the input file
name. For example, the first command below copies only the \s-1EVENT\s0 table,
while the second command copies other extensions as well:
.PP
.Vb 2
\&  [sh] funtable "/proj/rd/data/snr.ev[EVENTS]" events.ev
\&  [sh] funtable "/proj/rd/data/snr.ev[EVENTS+]" eventsandmore.ev
.Ve
.PP
If the input file is an image or a raw array file, then
\&\fBfuntable\fR will generate a \s-1FITS\s0 binary table from the pixel
values in the image. Note that it is not possible to specify the
columns to output (using command-line argument 3). Instead, there are
two ways to create such a binary table from an image. By default, a
3\-column table is generated, where the columns are \*(L"X\*(R", \*(L"Y\*(R", and
\&\*(L"\s-1VALUE\s0\*(R". For each pixel in the image, a single row (event) is
generated with the \*(L"X\*(R" and \*(L"Y\*(R" columns assigned the dim1 and dim2
values of the image pixel, respectively and the \*(L"\s-1VALUE\s0\*(R" column
assigned the value of the pixel. With sort of table, running
\&\fBfunhist\fR on the \*(L"\s-1VALUE\s0\*(R" column will give the same results as
running \fBfunhist\fR on the original image.
.PP
If the \fB\-i\fR (\*(L"individual\*(R" rows) switch is specified, then only
the \*(L"X\*(R" and \*(L"Y\*(R" columns are generated. In this case, each positive
pixel value in the image generates n rows (events), where n is equal
to the integerized value of that pixel (plus 0.5, for floating point
data).  In effect, \fB\-i\fR approximately recreates the rows of a
table that would have been binned into the input image. (Of course,
this is only approximately correct, since the resulting x,y positions
are integerized.)
.PP
If the \fB\-s [col1 col2 ... coln]\fR (\*(L"sort\*(R") switch is specified,
the output rows of a binary table will be sorted using the
specified columns as sort keys. The sort keys must be scalar columns
and also must be part of the output file (i.e. you cannot sort on a
column but not include it in the output). This facility uses the
\&\fB_sort\fR program (included with funtools), which must be accessible
via your path.
.PP
For binary tables, the \fB\-m\fR (\*(L"multiple files\*(R") switch will
generate a separate file for each region in the filter specification
i.e. each file contains only the rows from that region. Rows
which pass the filter but are not in any region also are put in a
separate file.
.PP
The separate output file names generated by the \fB\-m\fR switch are
produced automatically from the root output file to contain the region id of
the associated region. (Note that region ids start at 1, so that the
file name associated with id 0 contains rows that pass the filter but
are not in any given region.) Output file names are generated as follows:
.IP "\(bu" 4
A \f(CW$n\fR specification can be used anywhere in the root file name (suitably
quoted to protect it from the shell) and will be expanded to be the id
number of the associated region. For example:
.Sp
.Vb 1
\&  funtable -m input.fits'[cir(512,512,1);cir(520,520,1)...]' 'foo.goo_$n.fits'
.Ve
.Sp
will generate files named foo.goo_0.fits (for rows not in any region but

utils/funtools/funtable.c
/*
 *	Copyright (c) 1999-2003 Smithsonian Astrophysical Observatory
 */

#include <funtoolsP.h>
#include <find.h>
#include <zprocess.h>
#include <chandra.h>

#define MAXROW 8192
static int maxrow=MAXROW;
static char macrobuf[SZ_LINE];

#define SORTPROG "_sort"

typedef struct rowstruct{
  int x, y;
  double value;
  int region;
  int n;
} *Row, RowRec;

extern char *optarg;
extern int optind;

#ifdef ANSI_FUNC
static char *
_funtableCB(char *name, void *client_data)
#else
static char *_funtableCB(name, client_data)
	char *name;
	void *client_data;
#endif
{
  int n=*(int *)client_data;
  if( !strcasecmp(name, "n") ){
    snprintf(macrobuf, SZ_LINE-1, "%d", n);
    return(macrobuf);
  }
  else{
    return NULL;
  }
}

#ifdef ANSI_FUNC
static void
usage (char *fname)
#else
static void usage(fname)
     char *fname;
#endif
{
  fprintf(stderr,
	  "usage: %s [-a] [-i|-z] [-m] [-s cols] iname oname [columns]\n",
	  fname);
  fprintf(stderr, "optional switches:\n");
  fprintf(stderr, "  -a  # append to existing output file as a table extension\n");
  fprintf(stderr, "  -i  # for image data, only generate X and Y columns\n");
  fprintf(stderr, "  -m  # for tables, write a separate file for each region\n");
  fprintf(stderr, "  -s 'col1 ...'  # columns on which to sort\n");
  fprintf(stderr, "  -z  # for image data, output zero-valued pixels\n");
  fprintf(stderr, "\n(version: %s)\n", FUN_VERSION);
  exit(1);
}

#ifdef ANSI_FUNC
int 
main (int argc, char **argv)
#else
int
main(argc, argv)
     int argc;
     char **argv;
#endif
{
  int c;
  int i, j, k, n;
  int args;
  int nfile;
  int got, put;
  int type, isize;
  int dim1, dim2, dims;
  int ichan, ochan, pid, status;
  int rtype, rmode, roffset;
  int r=0;
  int rval=0;
  int ncol=0;
  int ip=0;
  int doindiv=0, dozero=0, dosort=0, domulti=0;
  int dtype, doffset, dmode, dn;
  int plen, elen, flen;
  int epad, ppad;
  double dval;
  char *iname=NULL;
  char *oname=NULL;
  char *mode=NULL;
  char *sortcols=NULL;
  char *actcols=NULL;
  char *prog=NULL;
  char *path=NULL;
  char *ebuf=NULL;
  char *eptr=NULL;
  char *omode="w";
  char *s=NULL;
  char *t=NULL;
  char *u=NULL;
  char *fits=NULL;
  char *extn=NULL;
  char col[SZ_LINE];
  char tbuf[SZ_LINE];
  char *pargs[SZ_LINE];
  double *dbuf=NULL;
  FILE *ifile;
  Fun fun, tfun=NULL;
  Fun *funs=NULL;
  Filter filt=NULL;
  FITSCard card;
  Row rowbuf, row;

  /* exit on gio errors */
  if( !getenv("GERROR")  )
    setgerror(2);

  /* we want the args in the same order in which they arrived, and
     gnu getopt sometimes changes things without this */
  putenv("POSIXLY_CORRECT=true");

  /* process switch arguments */
  while ((c = getopt(argc, argv, "aims:z")) != -1){
    switch(c){
    case 'a':
      omode = "A";
    case 'i':
      doindiv = 1;
      break;
    case 'm':
      domulti=1;
      break;
    case 's':
      dosort = 1;
      sortcols=xstrdup(optarg);
      break;
    case 'z':
      dozero = 1;
      break;
    }
  }

  /* get maxrow,if user-specified */
  if( (s=getenv("FUN_MAXROW")) != NULL )
    maxrow = atoi(s);

  /* check for required arguments */
  args = argc - optind;
  if( args < 2 ) usage(argv[0]);
  iname = argv[optind+0];
  oname = xstrdup(argv[optind+1]);

  if( !(fun = FunOpen(iname, "rc", NULL)) )
    gerror(stderr, "can't FunOpen input file (or find extension): %s\n",
	   iname);

  /* for multiple outputs, we need a file for each region and one extra */
  if( domulti ){
    if( fun->filter )
      tfun = fun;
    else{
      /* for non-disk files, we should punt, but we'll open a fake
	 Chandra/ACIS FITS file and hope for the best */
      if( !strncasecmp(iname, "stdin", 5) || !strncmp(iname, "-", 1) ||
	  !strncasecmp(iname, "pipe:", 5) ){
	extn = FileExtension(iname);
	plen = strlen(_prim);
	if( (ppad = FT_BLOCK - (plen % FT_BLOCK)) == FT_BLOCK ) ppad = 0;
	elen = strlen(_extn);
	if( (epad = FT_BLOCK - (elen % FT_BLOCK)) == FT_BLOCK ) epad = 0;
	flen = plen + ppad + elen + epad;
	fits = (char *)calloc(flen+1, sizeof(char));
	memset(fits, (int)' ', flen);
	memcpy(fits, _prim, plen);
	memcpy(fits+plen+ppad, _extn, elen);
	snprintf(tbuf, SZ_LINE-1,
		 "buf:%#lx:%d%s", (long)fits, flen, extn?extn:"");
	tfun = FunOpen(tbuf, "r", NULL);
      }
      /* otherwise we can open the input file once again */
      else
	tfun = FunOpen(iname, "r", NULL);
    }
    FunInfoGet(tfun, FUN_TYPE, &type, 0);
    if( (type == FUN_IMAGE) || (type == FUN_ARRAY) )
      gerror(stderr, "can't output to multiple image files\n");
    strcpy(tbuf, "lexonly=true");
    if( tfun->bincols ){
      strcat(tbuf, ",");
      strcat(tbuf, tfun->bincols);
    }
    filt = FilterOpen(tfun->header, tfun->filter, tbuf);
    nfile = FilterRegionCount(TOK_IREG)+1;
    FilterClose(filt);
    if( tfun != fun )
      FunClose(tfun);
  }
  /* output a single file */
  else{
    nfile = 1;
    /* current fun handle is always first */
    r = 0;
  }
  /* should work for multi and single file output */
  if( !(funs = (Fun *)xcalloc(nfile, sizeof(Fun))) )
    gerror(stderr, "can't allocate space for multiple fun handles\n");
  /* make up output file names and open files */
  if( domulti ){
    /* open the input FITS file */
    if( !(u=strchr(oname, '$')) ){
      s = calloc(SZ_LINE, sizeof(char));
      if( (t=strchr(oname, '.')) ){
	*t='\0';
	t++;
      }
    }
    for(i=0; i<nfile; i++){
      /* put ivalue exactly where the user asked for it */
      if( u )
	s = ExpandMacro(oname, NULL, NULL, 0, _funtableCB, (void *)&i);
      /* put ivalue before first '.' */
      else if( t )
	snprintf(s, SZ_LINE-1, "%s%d.%s", oname, i, t);
      /* no hint about what to do -- prepend the ivalue */
      else
	snprintf(s, SZ_LINE-1, "%s%d", oname, i);
      /* last one triggers open of input file and possible copy */
      if( i == (nfile-1) )
	strncpy(tbuf, omode, SZ_LINE-1);
      /* otherwise, just add output fd to list of output file */
      else
	snprintf(tbuf, SZ_LINE-1, "%sd", omode);
      /* open the output FITS table, preparing to copy input params.
	 must be done before accessing the input file to allow for ext copy */
      if( !(funs[i] = FunOpen(s, tbuf, fun)) )
	gerror(stderr, "can't FunOpen output file #%d: %s\n", i, s);
      if( u && s ) xfree(s);
    }
    if( !u && s ) xfree(s);
  }
  else{
    if( !(funs[r] = FunOpen(oname, omode, fun)) )
      gerror(stderr, "can't FunOpen output file: %s\n", oname);
  }

  /* get type of input data */
  FunInfoGet(fun, FUN_TYPE, &type, 0);

  /* process data by type */
  switch(type){
  case FUN_TABLE:
  case FUN_EVENTS:
    /* uness we are sorting, only read $region & $n into user space so we
       can override their values (by merging with raw data) */
    if( !dosort ){
      FunColumnSelect(fun, sizeof(RowRec), "merge=replace",
		      "$REGION", "J", "rw", FUN_OFFSET(Row, region),
		      "$N",      "J", "rw", FUN_OFFSET(Row, n),
		      NULL);
      FunColumnActivate(fun, "-$REGION -$N", NULL);
    }

    /* set activate specified output columns */
    if( args >= 3 ) actcols = argv[optind+2];
    FunColumnActivate(fun, actcols, NULL);

    /* if region was not activiated by the user, activate now for read only */
    if( domulti ){
      if( !FunColumnLookup(fun, COL_REGION_NAME,
			   0, NULL, &rtype, &rmode, &roffset, NULL, NULL) ||
	  !(rmode&COL_ACTIVE))
      FunColumnActivate(fun, "+$REGION", "mode=r");
    }
    /* if we specified columns, we also can specify "mask=transparent"
       and get all events, with the regionid (assuming we specified 
       the region column to display as well) */
    if( argc >= 4 )
      mode = argv[optind+3];
    /* set up sorting, if necessary */
    if( dosort ){
      /* look for sort program */
      path = (char *)getenv("PATH");
      if( !(prog=Find(SORTPROG, "x", NULL, path)) )
	gerror(stderr, "can't locate sort program: %s\n", SORTPROG);
      /* get the size of the input record */
      FunInfoGet(fun, FUN_ROWSIZE, &isize, 0);
      /* construct argument list for sort program */
      pargs[0] = (char *)prog;
      snprintf(tbuf, SZ_LINE-1, "-B%d", isize);
      pargs[1] = (char *)xstrdup(tbuf);
      for(ncol=0; word(sortcols, col, &ip); ncol++ ){
	if( !FunColumnLookup(fun, col, 0, NULL, &dtype, &dmode, &doffset, &dn,
			     NULL) ){
	  gerror(stderr, "can't find sort column '%s'\n", col);
	}
	if( !(dmode & COL_ACTIVE) ){
	  gerror(stderr, "sort column '%s' must be active\n", col);
	}
	switch(dtype){
	case 'A':
	  snprintf(tbuf, SZ_LINE-1, "+t%d.%d", doffset, dn);
	  break;
	case 'B':
	  snprintf(tbuf, SZ_LINE-1, "+B%d", doffset);
	  break;
	case 'I':
	  snprintf(tbuf, SZ_LINE-1, "+s%d", doffset);
	  break;
	case 'U':
	  snprintf(tbuf, SZ_LINE-1, "+S%d", doffset);
	  break;
	case 'J':
	  snprintf(tbuf, SZ_LINE-1, "+i%d", doffset);
	  break;
	case 'K':
	  snprintf(tbuf, SZ_LINE-1, "+l%d", doffset);
	  break;
	case 'V':
	  snprintf(tbuf, SZ_LINE-1, "+I%d", doffset);
	  break;
	case 'E':
	  snprintf(tbuf, SZ_LINE-1, "+f%d", doffset);
	  break;
	case 'D':
	  snprintf(tbuf, SZ_LINE-1, "+d%d", doffset);
	  break;
	default:
	  gerror(stderr, "unsupported sort data type for column %s\n", tbuf);
	}
	pargs[ncol+2] = (char *)xstrdup(tbuf);
	if( ncol > (SZ_LINE-3) ){
	  gerror(stderr, "ERROR: too many sort column arguments: %d\n", ncol);
	}
      }
      /* null terminate arg list */
      pargs[ncol+2] = NULL;
      /* start the sort program */
      if( !ProcessOpen(prog, pargs, &ichan, &ochan, &pid) ){
	gerror(stderr, "ERROR: can't start sort program: %s\n", prog);
      }
    }
    /* extract events */
    while( (ebuf = FunTableRowGet(fun, NULL, maxrow, mode, &got)) ){
      /* sort: write unsorted events to sort program */
      if( dosort ){
	write(ochan, ebuf, got*isize);
      }
      /* no sort: copy directly to output */
      else{
	/* multiple files -- one for each region id */
	if( domulti ){
	  for(n=0; n<got; n++){
	    row = (Row)ebuf+n;
	    r = MAX(0,row->region);
	    if( (put=FunTableRowPut(funs[r], row, 1, n, NULL)) != 1 ){
	      gerror(stderr, "expected to write %d rows; only wrote %d\n",
		     1, put);
	    }
	  }
	}
	/* single file output */
	else{
	  if( (put=FunTableRowPut(funs[r], ebuf, got, 0, NULL)) != got ){
	    gerror(stderr, "expected to write %d rows; only wrote %d\n",
		   got, put);
	  }
	}
      }
      xfree(ebuf);
    }
    /* finish up sorting, if necessary */
    if( dosort ){
      /* signal EOF to sort program */
      close(ochan);
      if( maxrow * isize > fun->maxbufsize ){
      if( isize > fun->maxbufsize )
	maxrow = 1;
      else
	maxrow = fun->maxbufsize / isize;
      }
      if( maxrow <= 0 ) maxrow = 1;
      ebuf = xmalloc(maxrow*isize);
      /* use stdio to avoid returning partial records */
      if( !(ifile = fdopen(ichan, "r")) ){
	gerror(stderr, "can't retrieve sorted records\n");
      }
      while( (got=fread(ebuf, isize, maxrow, ifile)) ){
	/* multiple files -- one for each region id */
	if( domulti ){
	  for(n=0; n<got; n++){
	    /* point at next event */
	    eptr = ebuf + (n*isize);
	    /* get region id */
	    switch(rtype){
	    case 'B':
	      rval = (int)*(unsigned char *)(eptr+roffset);
	      break;
	    case 'I':
	      rval = (int)*(short *)(eptr+roffset);
	      break;
	    case 'U':
	      rval = (int)*(unsigned short *)(eptr+roffset);
	      break;
	    case 'J':
	      rval = *(int *)(eptr+roffset);
	      break;
	    case 'K':
#if HAVE_LONG_LONG == 0
	      gerror(stderr, 
		  "64-bit data support not built (long long not available)\n");
#endif
	      rval = *(longlong *)(eptr+roffset);
	      break;
	    case 'V':
	      rval = (int)*(unsigned int *)(eptr+roffset);
	      break;
	    case 'E':
	      rval = (int)*(float *)(eptr+roffset);
	      break;
	    case 'D':
	      rval = (int)*(double *)(eptr+roffset);
	      break;
	    }
	    /* offset is region id (but not less than 0) */
	    r = MAX(0, rval);
	    /* output event to appropriate file */
	    if( (put=FunTableRowPut(funs[r], eptr, 1, n, NULL)) != 1 ){
	      gerror(stderr, "expected to write %d rows; only wrote %d\n",
		     1, put);
	    }
	  }
	}
	/* single file output */
	else{
	  /* output all events in this buffer to single file */
	  if( (put=FunTableRowPut(funs[r], ebuf, got, 0, NULL)) != got ){
	    gerror(stderr, "expected to write %d rows; only wrote %d\n",
		   got, put);
	  }
	}
      }
      /* done with sorting process */
      ProcessClose(pid, &status);
      fclose(ifile);
      if( sortcols ) xfree(sortcols);
      if( ebuf ) xfree(ebuf);
      for(i=0; i<ncol+2; i++)
	if( pargs[i] ) xfree(pargs[i]);
    }
    /* write out heap space, if it exists -- this quick hack needs cleanup! */
    if( fun->header && (ft_pcount(fun->header) > 0) ){
      int got;
      int hlen;
      char *heap;
      hlen=ft_pcount(fun->header);
      if( (heap = xmalloc(hlen)) ){
	/* read heap data */
	got = gread (fun->gio, heap, sizeof(char), hlen);
	fun->bytes += got;
	/* write heap data to each output file */
	for(n=0; n<nfile; n++){
	  got = gwrite (funs[n]->gio, heap, sizeof(char), got);
	  funs[n]->bytes += got;
	}
	xfree(heap);
      }
    }
    break;
  case FUN_IMAGE:
  case FUN_ARRAY:
    if( dosort )  gerror(stderr, "can't sort image data\n");
    if( domulti ) gerror(stderr, "can't output to multiple image files\n");
    /* get image section information */
    FunInfoGet(fun, 
	       FUN_SECT_DIM1, &dim1,
	       FUN_SECT_DIM2, &dim2,
	       FUN_SECT_DIMS, &dims, 0);
    /* make a standard extension name for the new binary table */
    FunParamPuts(funs[r], "EXTNAME", 0, "EVENTS", NULL, 1);
    if( dims >= 2 ){
      /* add tlmin/tlmax values for the x,y columns */
      FunParamPuti(funs[r], "TLMIN1", 0, 1, NULL, 1);
      FunParamPuti(funs[r], "TLMAX1", 0, dim1, NULL, 1);
      FunParamPuti(funs[r], "TLMIN2", 0, 1, NULL, 1);
      FunParamPuti(funs[r], "TLMAX2", 0, dim2, NULL, 1);
      /* transform WCS cards from image to binary table */
      for(i=1; i<=2; i++){
	if( (s=ft_headgets(fun->header, "CTYPE", i, NULL, &card)) && card ){
	  FunParamPuts(funs[r], "TCTYP", i, s, NULL, 1);
	  xfree(s);
	}
	if( (s=ft_headgets(fun->header, "CRVAL", i, NULL, &card)) && card ){
	  FunParamPuts(funs[r], "TCRVL", i, s, NULL, 1);
	  xfree(s);
	}
	if((dval=ft_headgetr(fun->header, "CDELT", i, 0.0, &card)) && card){
	  FunParamPutd(funs[r], "TCDLT", i, dval, 7, NULL, 1);
	}
	if( (s=ft_headgets(fun->header, "CROTA", i, NULL, &card)) && card ){
	  FunParamPuts(funs[r], "TCROT", i, s, NULL, 1);
	  xfree(s);
	}
	if((dval=ft_headgetr(fun->header, "CRPIX", i, 0.0, &card)) && card){
	  FunParamPutd(funs[r], "TCRPX", i, dval, 7, NULL, 1);
	}
      }
      /* select columns */
      if( doindiv ){
	FunColumnSelect(funs[r], sizeof(RowRec), NULL,
			"X",       "J",  "w",  FUN_OFFSET(Row, x),
			"Y",       "J",  "w",  FUN_OFFSET(Row, y),
			NULL);
      }
      else{
	FunColumnSelect(funs[r], sizeof(RowRec), NULL,
			"X",       "J",  "w",  FUN_OFFSET(Row, x),
			"Y",       "J",  "w",  FUN_OFFSET(Row, y),
			"VALUE",   "D",  "rw", FUN_OFFSET(Row, value),
			NULL);
      }
    }
    else{
      /* set dim2 to 1 for 1D image */
      dim2 = 1;
      /* select columns */
      if( doindiv ){
	FunColumnSelect(funs[r], sizeof(RowRec), NULL,
			"X",       "J",  "w",  FUN_OFFSET(Row, x),
			NULL);
      }
      else{
	FunColumnSelect(funs[r], sizeof(RowRec), NULL,
			"X",       "J",  "w",  FUN_OFFSET(Row, x),
			"VALUE",   "D",  "rw", FUN_OFFSET(Row, value),
			NULL);
      }
    }
    /* we will make event row(s) for each x,y pixel */
    rowbuf = xcalloc(dim1, sizeof(RowRec));
    /* process each image row */
    for(j=1; j<=dim2; j++){
      /* extract image data and make event rows */
      if( !(dbuf = FunImageRowGet(fun, dbuf, j, j, "bitpix=-64")) )
	gerror(stderr, "can't FunImageRowGet: %s\n", iname);
      if( doindiv ){
	/* generate an event for this pixel */
	for(i=0; i<dim1; i++){
	  /* replicate individual row according to the value of the pixel.
	     don't replicate pixels values less than zero; round fp values */
	  if( dbuf[i] >0 ){
	    rowbuf[i].x = i+1;
	    rowbuf[i].y = j;
	    for(k=0; k<(int)(dbuf[i]+0.5); k++){
	      /* write replicated row */
	      if( FunTableRowPut(funs[r], &rowbuf[i], 1, 0, NULL) != 1 )
		gerror(stderr, "can't FunTableRowPut: %s\n", oname);
	    }
	  }
	}
      }
      else{
	/* generate one event per pixel, with an accompanying value column */
	for(i=0; i<dim1; i++){
	  /* don't write zero rows unless asked */
	  if( dozero || (dbuf[i] > 0) ){
	    rowbuf[i].x = i+1;
	    rowbuf[i].y = j;
	    rowbuf[i].value = dbuf[i];
	    /* write row */
	    if( FunTableRowPut(funs[r], &rowbuf[i], 1, 0, NULL) != 1 )
	      gerror(stderr, "can't FunTableRowPut: %s\n", oname);
	  }
	}
      }
    }
    if( rowbuf ) xfree(rowbuf);
    if( dbuf ) xfree(dbuf);
    break;
  default:
    gerror(stderr, "unknown Funtools data type: %d\n", type);
    exit(1);
  }

  /* for multiple outputs, flush the input extension to all files at once */
  if( domulti ){
    FunFlush(fun, "copy=ref");
  }

  /* if we close output before input, all flushing is done automatically  */
  /* free up space from multiple fun handles */
  for(i=0; i<nfile; i++){
    if( funs[i] ){
      FunClose(funs[i]);
    }
  }
  xfree(funs);
  if( oname ) xfree(oname);
  if( fits ) xfree(fits);
  if( extn ) xfree(extn);
  FunClose(fun);
  return(0);
}
